core/src/main/java/org/apache/calcite/rel/rules/JoinConditionOrExpansionRule.java

```diff
@@ -55,40 +62,92 @@ protected JoinConditionOrExpansionRule(Config config) {
   @Override public void onMatch(RelOptRuleCall call) {
     Join join = call.rel(0);
     RelBuilder relBuilder = call.builder();
-    List<RexNode> orConds = RelOptUtil.disjunctions(join.getCondition());
-
-    if (orConds.size() <= 1) {
-      return;
-    }
 
     RelNode converted;
     switch (join.getJoinType()) {
     case INNER:
-      converted = expandInnerJoin(join, orConds, relBuilder);
-      break;
-    case SEMI:
-      converted =
-          expandSemiOrAntiJoin(join, JoinRelType.SEMI, orConds, relBuilder);
+      converted = expandInnerJoin(join, relBuilder);
       break;
     case ANTI:
-      converted =
-          expandSemiOrAntiJoin(join, JoinRelType.ANTI, orConds, relBuilder);
+      converted = expandAntiJoin(join, relBuilder);
       break;
     case LEFT:
-      converted = expandLeftJoin(join, orConds, relBuilder);
+      converted = expandLeftOrRightJoin(join, true, relBuilder);
       break;
     case RIGHT:
-      converted = expandRightJoin(join, orConds, relBuilder);
+      converted = expandLeftOrRightJoin(join, false, relBuilder);
       break;
     case FULL:
-      converted = expandFullJoin(join, orConds, relBuilder);
+      converted = expandFullJoin(join, relBuilder);
       break;
     default:
       return;
     }
+
+    if (converted == null) {
+      return;
+    }
     call.transformTo(converted);
   }
 
+  private boolean skip(Join join) {
+    List<RexNode> orConds = RelOptUtil.disjunctions(join.getCondition());
+    return orConds.size() <= 1;
+  }
+
+  private List<RexNode> splitCond(Join join) {
+    RexBuilder builder = join.getCluster().getRexBuilder();
+    List<RexNode> conds = new ArrayList<>();
+    List<RexNode> otherConds = new ArrayList<>();
+    List<RexNode> orConds = RelOptUtil.disjunctions(join.getCondition());
+    int leftFieldCount = join.getLeft().getRowType().getFieldCount();
+
+    for (RexNode cond : orConds) {
+      if (!isUsefulCond(cond, leftFieldCount)) {
```

Comment:
So you are not preserving the order...
This may cause problems with operations that cause runtime exceptions. 
Note that almost anything can cause runtime exceptions, e.g., arithmetic overflow.

Reply:
If you want to be conservative you have to:
- either detect whether an expression can cause a runtime exception and only move the ones that never do
- preserve the order and stop at the first expression which is not useful and put everything from that point on in a single big OR which is not expanded

Can you add a test for such a case so we can see what happens?
For example an expression such as `A <= 0 OR LN(A) <= 0`

Reply:
I kept the order of the join conditions and also fixed a few details you mentioned. Thanks so much!